### ISSUE TITLE
Multi-view consistency

### ISSUE THREAD

**Issue by EchoTHChen** [AugmentariumLab/omnisyn#6]
I met a new problem about the predicted depthmap of DepthNet in OmniSyn. I run the depth estimation network to predict depthmaps for two panoramas of one scene in Matterport3d (The baseline is 1.0 meter).

I get the 3D point clouds according to the depthmaps of the two panoramas. But two surface appeared in the 3D pointclouds finally. It seems that it is actually not view-consistent for the predicted depthmaps. But the cost-volume has been used in the code. I do not know why it occurs. Could you give me some suggestions for resolving the issue?
<img width="695" alt="snapshot01" src="https://user-images.githubusercontent.com/32491440/183370646-e57102b6-3ffe-48ef-a7b5-e21f6858b3d5.png">

<img width="695" alt="snapshot02" src="https://user-images.githubusercontent.com/32491440/183370786-cd23a85e-5a99-4040-81c3-5614fd07f7f2.png">

**Comment by dli7319** [comment]
Hello EchoTHChen,

A bug in the rectify function in `models/pipeline3_model.py` (d747478e2a) caused the cost volume to not work for one of the panos.
I think this is what caused the issue. 
That said, I don't think the depth even with this fixed is good enough for 3D modeling so keep that in mind.

I also added some point cloud visualization code which you can try with
```
python train_inpainting.py -c example/config_example_m3d.txt
```

Point Cloud with GT depth:
![Screenshot from 2022-08-10 13-10-15](https://user-images.githubusercontent.com/27523415/183977142-9d2f13b1-edf8-44b7-a6b5-7017ad3b46bf.png)
Point Cloud w/ predicted depth before bug fix:
![Screenshot from 2022-08-10 13-08-02](https://user-images.githubusercontent.com/27523415/183977180-683d09e6-909b-49d2-9be6-1b2d90416af4.png)
Point Cloud w/ predicted depth after bug fix:
![Screenshot from 2022-08-10 13-08-47](https://user-images.githubusercontent.com/27523415/183977213-0f47e1f1-5b4b-43ac-ad14-b74e3b0da488.png)

**Comment by EchoTHChen** [comment]
It seems that you only revised the code in Line 1968 rectify_images function.  Do I need to retrain the depth estimation network from scatch? Or I just use the pretrained model(before you fix the bug) to re-inference with the revised code?

**Comment by dli7319** [comment]
You don't need to retrain.

**Comment by EchoTHChen** [comment]
I'm new to panorama depth esimation. Why rectify_images is needed?

**Comment by dli7319** [comment]
For the cost volume, I need to project pixels from one panorama to another. For this you need the pose of each panorama. 

In my code, I decided to first rotate the panoramas so I know one panorama is in front of another so it is easier to code the cost volume portion. In retrospect, it's probably better to just multiply each pixel through the transformation matrix for each panorama.
When that function was not working, the projection is correct for one image but backward for the second so the (deep features) cost volume did not line up properly when I swapped the two images:

correct cost volume visualization using RGB values:
![cv](https://user-images.githubusercontent.com/27523415/184056537-5587ddfa-15f3-4d03-a30b-33cab47fe1c7.gif)

incorrect rectification:
![cv_bad](https://user-images.githubusercontent.com/27523415/184056590-f5f60a65-04c7-4d80-aeed-c167f950dc04.gif)

**Comment by EchoTHChen** [comment]
Thanks

**Comment by EchoTHChen** [comment]
I came across another issue after I use the updated code to generate the point clouds:
The given code generates normal point clouds for the scene in the validation set (The visualize_point_cloud function is tested in the validation set data). But it fails to generate the accurate point clouds in the test dataset (the test part of HabitatImageGenerator in run_inpainting_single function):

![image](https://user-images.githubusercontent.com/32491440/184595082-e618cd71-5298-4961-8605-d6b3728c2c9f.png)

, which is worse than the result of the previous code.


The Ground Truth is as follows:
![image](https://user-images.githubusercontent.com/32491440/184595583-458d804d-c336-4a59-8580-e2b13bbe1466.png)

Do you know how to fix this bug?

**Comment by dli7319** [comment]
Hey there, 

I'm having trouble reproducing the issue.
Can you please provide the code you used to get that?
Also, are you using my weights or your own trained weights? 

This is what I get using this code in the visualization_point_cloud function which I think gets the same scene as what you have:
```python
# After line 1673
    seq_len = 3
    reference_idx = 1
    test_data = HabitatImageGenerator(
      "test",
      full_width=self.full_width,
      full_height=self.full_height,
      seq_len=seq_len,
      reference_idx=reference_idx,
      m3d_dist=1.0
    )

    val_dataloader = DataLoader(test_data,
                            batch_size=1,
                            shuffle=False,
                            num_workers=0 if args.dataset == 'm3d' else 4)
```

Predicted:
![Screenshot from 2022-08-15 04-15-12](https://user-images.githubusercontent.com/27523415/184600606-8c07f396-6808-4643-a69f-8068dbdfb960.png)
![Screenshot from 2022-08-15 04-15-14](https://user-images.githubusercontent.com/27523415/184600632-7323c9a0-ae46-49c5-94d9-3ef8ac1caea7.png)

GT:
![Screenshot from 2022-08-15 04-16-08](https://user-images.githubusercontent.com/27523415/184600650-243a1667-4db0-4f1f-a271-92748f7d3d37.png)

**Comment by EchoTHChen** [comment]
Now, I generate almost the same point clouds as yours by using the given visulize_point_cloud function. But I find that I fail to produce normal inpainted results by running `python train_inpainting.py --config configs/m3d_inpainting_config.txt \
    --use-pred-depth True \
    --checkpoints-dir runs/run_m3d_depth \
    --cost-volume v3_erp \
    --depth-input-uv True \
    --model-use-v-input True \
    --script-mode run_inpainting_single`

![1](https://user-images.githubusercontent.com/32491440/184608541-3c889e3c-6ef4-4ec3-9a38-e02f6a0a5b8c.png)
![2](https://user-images.githubusercontent.com/32491440/184608555-ce59a850-c099-4a32-9f9c-680407f32ac4.png)
![3](https://user-images.githubusercontent.com/32491440/184608567-a98b403a-ca64-4021-9c40-d5eee9107e18.png)

**Comment by EchoTHChen** [comment]
I think there are some problems in the run_inpaint_single function even given a normal depth estimation.

**Comment by EchoTHChen** [comment]
I'm using my own trained weights.

**Comment by dli7319** [comment]
Does it work for you using the example config and weights?
```
python train_inpainting.py --config example/config_example_m3d.txt \
    --use-pred-depth True \
    --script-mode run_inpainting_single
```

If so, maybe you can check your trained depth model w/ our inpainting model or your inpainting w/ our depth to figure out which one isn't working?

It should look like this:
![2](https://user-images.githubusercontent.com/27523415/184613938-4ba82886-f618-499a-b287-c430704b8a7e.gif)

**Comment by EchoTHChen** [comment]
The example config works. Thanks!



---Original---
From: "David ***@***.***&gt;
Date: Mon, Aug 15, 2022 17:36 PM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [AugmentariumLab/omnisyn] Multi-view consistency (Issue #6)




 
Does it work for you using the example config and weights?
 python train_inpainting.py --config example/config_example_m3d.txt \     --use-pred-depth True \     --script-mode run_inpainting_single  
If so, maybe you can check your trained depth model w/ our inpainting model or your inpainting w/ our depth to figure out which one isn't working?
